packages/compartment-mapper/src/bundle.js

```diff
@@ -17,21 +17,23 @@ import { compartmentMapForNodeModules } from './node-modules.js';
 import { search } from './search.js';
 import { link } from './link.js';
 import { makeImportHookMaker } from './import-hook.js';
-import { parseJson } from './parse-json.js';
-import { parseArchiveCjs } from './parse-archive-cjs.js';
-import { parseArchiveMjs } from './parse-archive-mjs.js';
+import parserJson from './parse-json.js';
+import parserArchiveCjs from './parse-archive-cjs.js';
+import parserArchiveMjs from './parse-archive-mjs.js';
 import { parseLocatedJson } from './json.js';
 
 const textEncoder = new TextEncoder();
 
 /** quotes strings */
 const q = JSON.stringify;
 
-/** @type {Record<string, ParseFn>} */
+/** @type {Record<string, ParserImplementation>} */
 const parserForLanguage = {
-  mjs: parseArchiveMjs,
-  cjs: parseArchiveCjs,
-  json: parseJson,
+  mjs: parserArchiveMjs,
```

Comment:
This is a little weird because an archive could theoretically contain a mix of mjs and pre-mjs-json, but we explicitly don’t allow `mjs` because to do so would necessitate a dependency on Babel in parts of our system where 3MB of JS is prohibitively expensive.

Is it possible to suffer the type narrowing here with more grace? I could accept stub implementations of the mjs and cjs parsers that just throw if those types are found in archives.

Reply:
**TL;DR:**

I didn't add the `mjs`, I only renamed the variable parse->parse_**r**_.  
I've added `pre-mjs-json`. And they're both pointing to the same parser reference. I didn't change the list in import-archive. I don't see how that adds new error potential.

That said, I'd like to refactor this to provide more strict typing between how parsers identify themselves and the lparserForLanguage/mapParsers code. 

My plan is to merge this (I find it safe to merge) and proceed to refactor.

----
**All of it:**

Well, the issue here is that `Language` is really 2 types.   

The language of a file as declared in package.json or extension: `Language1` := mjs, cjs, json   
The `parser` field meaning "which parser produced this": `Language2` := `Language1` + pre-mjs-json, pre-cjs-json  

parserForLanguage maps from `Language1` to a parser that's supposed to be used for processing whatever's loaded in the current usecase (therefore it differs between load, bundle and archive)

Meanwhile the lookup to check if the parser is using a heuristic to come up with imports is only given access to `Language2` typed `parser` field.

I had the choice to:
- follow the `Language` type definition and fill the maps with missing items
- add a few lines in link.js that, instead of using the value that came from the parts of the implementation that understands esm/csj nuance, is coding the logic to understnad that  pre-mjs-json maps to mjs etc.

Allowing link.js to be free of cjs/esm knowledge seemed a better choice.

> but we explicitly don’t allow mjs
Note that the fields that I'm adding are `pre-*` and mjs was already there. Newly added fields refer to the same parser implementation as their original counterparts and the actual value of the parser field is `pre-*`
I did not change the keys in import-archive and the bundle implementation doesn't use the newly added fields when looking up to choose a parser. Ideally the `parser` field in the `parse` function output would correspond to the key under which it's found in the parserForLanguage map. 

I'd see it wrapped in a language util (not the final name hopefully) which is fed applicable parsers and returns the parserForLanguage map. I'd go ahead and implement it right away but I don't fully understand the concern of `mjs` showing up in an archive and what's the intended reaction if it's there. My mental model tells me it'd go to import-archive and fail there and it's not a concern for the archive maker. My mental model doesn't fully cover bundling though.

Overall I think this implementation doesn't open new paths to language format errors that didn't exist before. I'd like to refactor the parser selection next so that we can, and I can't believe it's _me_ saying that, strict type matching between what parser says it is and the parserForLanguage map somehow. 


Possible designs:

### A:
Explicitly define 2 types for Language and resolve the type mismatch where it manifests itself, add a utility to handle creating parserForLanguage and/or mapping between Language1 and Language2 where needed

### B:
Drop the declarative mapping from languages to parsers and instead let parser implementations provide a function isResponsibleFor(Language): boolean, make parsers a list and iterate the list calling the isResponsibleFor instead of lookup. With 3 parsers and not more than 5 in reasonably foreseeable future the performance impact is none and the responsibility in load/archive/bundle is limited to providing the right set of parsers

I'd go for B. *Debate me* :D

Reply:
I grabbed this branch and removed just `pre-mjs-json` and `pre-cjs-json` in `bundle.js` and nothing broke. But, I now see the problem with removing these in `archive.js`. I’m not sure what solution I favor at this point because I don’t fully understand the bug in what I wrote, but I suspect that this should be solved by dividing the language type.

My vision for language parsers is to eventually make it possible to have `package.json` annotations to allow a per-package mapping from extension to language, and languages to custom parser implementations. So, the key type for this really ought to remain quite liberal (`string`) and we ought generally to check whether an entry is present anywhere we’re consulting it.

Reply:
> removed just pre-mjs-json and pre-cjs-json in bundle.js and nothing broke

IMHO it's more likely to mean I need to write more tests for bundle.js

>  quite liberal

Ok, so refactoring option B is off the table. That was a short debate.